sqlobject/src/main/java/org/jdbi/v3/sqlobject/internal/SqlObjectInitData.java

```diff
@@ -0,0 +1,134 @@
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.function.Supplier;
+import java.util.function.UnaryOperator;
+
+import org.jdbi.v3.core.config.ConfigRegistry;
+import org.jdbi.v3.core.extension.ExtensionMethod;
+import org.jdbi.v3.core.extension.HandleSupplier;
+import org.jdbi.v3.core.internal.exceptions.Sneaky;
+import org.jdbi.v3.sqlobject.GenerateSqlObject;
+import org.jdbi.v3.sqlobject.Handler;
+
+public final class SqlObjectInitData {
+    private static final Object[] NO_ARGS = new Object[0];
+    public static final ThreadLocal<SqlObjectInitData> INIT_DATA = new ThreadLocal<>();
+
+    public final boolean concrete;
+    public final Class<?> extensionType;
+    public final UnaryOperator<ConfigRegistry> instanceConfigurer;
+    public final Map<Method, UnaryOperator<ConfigRegistry>> methodConfigurers;
+    public final Map<Method, Handler> methodHandlers;
+
+    public SqlObjectInitData(
+            Class<?> extensionType,
+            UnaryOperator<ConfigRegistry> instanceConfigurer,
+            Map<Method, UnaryOperator<ConfigRegistry>> methodConfigurers,
+            Map<Method, Handler> methodHandlers) {
+        concrete = isConcrete(extensionType);
+        this.extensionType = extensionType;
+        this.instanceConfigurer = instanceConfigurer;
+        this.methodConfigurers = methodConfigurers;
+        this.methodHandlers = methodHandlers;
+    }
+
+    public static boolean isConcrete(Class<?> extensionType) {
+        return extensionType.getAnnotation(GenerateSqlObject.class) != null;
+    }
+
+    public static SqlObjectInitData initData() {
+        final SqlObjectInitData result = INIT_DATA.get();
+        if (result == null) {
+            throw new IllegalStateException("Implemented SqlObject types must be initialized by SqlObjectFactory");
+        }
+        return result;
+    }
+
+    public static Method lookupMethod(String methodName, Class<?>... parameterTypes) {
+        return lookup0(initData().extensionType, methodName, parameterTypes);
+    }
+
+    private static Method lookup0(Class<?> klass, String methodName, Class<?>... parameterTypes) {
+        try {
+            return klass.getMethod(methodName, parameterTypes);
+        } catch (NoSuchMethodException | SecurityException e) {
+            try {
+                return klass.getDeclaredMethod(methodName, parameterTypes);
+            } catch (Exception x) {
+                e.addSuppressed(x);
+            }
+            throw new IllegalStateException(
+                    String.format("can't find %s#%s%s", klass.getName(), methodName, Arrays.asList(parameterTypes)), e);
+        }
+    }
+
+    // Thanks Holger!
+    // https://stackoverflow.com/questions/35331327/does-java-8-have-cached-support-for-suppliers
+    public Supplier<InContextInvoker> lazyInvoker(Object target, Method method, HandleSupplier handle, ConfigRegistry instanceConfig) {
+        return new Supplier<InContextInvoker>() {
```

Comment:
I think at this point it's better to just make this a proper class of its own, this is more than just a lambda

Reply:
The argument capture is too nice here.  Writing out the constructor and fields and copy statements brings no value.

Reply:
I'm talking about the `LazyInvoker` part, not the rest of the body yet. The LazyInvoker contraption returned by this method is definitely an independent component of which the complexity can be set apart from all this.